### ISSUE TITLE
Documenter fails when using functional endpoints (i.e. RouterFunction beans)

### ISSUE THREAD

**Issue by barthand** [odrotbohm/moduliths#146]
Hello, 

It seems `moduliths-docs` doesn't support projects leveraging functional web endpoints registration - be it either WebMvc.fn or WebFlux.fn (fails for both, checked myself). 

Project setup:
Java 11 + Kotlin (checked also Java 8)
Spring Boot 2.4.1 (checked also using Spring Boot 2.3.7)
Moduliths 1.1.0-M1 (checked also [current main branch](https://github.com/odrotbohm/moduliths/commit/b46966565975df00ad9fcbaec558d9b7ce9bd0a6))

Test body (in Kotlin):
```
class DocumentingTest {

    @Test
    @Throws(IOException::class)
    fun verifyModularity() {
        val modules: Modules = Modules.of(ModulithsDemoApplication::class.java)

        modules.verify()

        // Generate documentation
        val documenter = Documenter(modules)
        val options: Options = Options.defaults() //
            .withExclusions { module -> module.getName().matches(".*core|.*support".toRegex()) }

        // Write overall diagram
        // FIXME This fails if RouterFunction is registered as a bean :(
        documenter.writeModulesAsPlantUml(options)
```

Stack-trace starting from the test-run:
```
java.lang.NullPointerException: Method may not be called before construction of hierarchy is complete

	at com.tngtech.archunit.thirdparty.com.google.common.base.Preconditions.checkNotNull(Preconditions.java:787)
	at com.tngtech.archunit.core.domain.JavaClass.getAllMethods(JavaClass.java:788)
	at org.moduliths.model.Module$ModuleDependency.fromMethodsOf(Module.java:523)
	at org.moduliths.model.Module$ModuleDependency.fromType(Module.java:499)
	at org.moduliths.model.Module.lambda$getDirectModuleDependencies$16(Module.java:361)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:271)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.moduliths.docs.Documenter.addDependencies(Documenter.java:266)
	at org.moduliths.docs.Documenter.lambda$getComponents$1(Documenter.java:126)
	at java.base/java.util.HashMap.forEach(HashMap.java:1336)
	at org.moduliths.docs.Documenter.getComponents(Documenter.java:126)
	at org.moduliths.docs.Documenter.lambda$addComponentsToView$13(Documenter.java:288)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1675)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.moduliths.docs.Documenter.addComponentsToView(Documenter.java:289)
	at org.moduliths.docs.Documenter.createPlantUml(Documenter.java:370)
	at org.moduliths.docs.Documenter.writeModulesAsPlantUml(Documenter.java:145)
	at com.example.modulithsdemo.DocumentingTest.verifyModularity(DocumentingTest.kt:28)
```

I tried adding to `org.springframework.web.servlet.function` and `org.springframework.web.reactive.function` to `org.moduliths.model.Modules#FRAMEWORK_PACKAGES`, but this causes some further NPEs while rendering AsciiDoctor.

P.S. Thanks a lot (!) for this project and efforts put here, I think this represents a wide gap in current Java toolset.

**Comment by odrotbohm** [comment]
Do you have a sample project available for me to play with? The stacktrace seems to hint at an ArchUnit problem actually but we're effectively using `JavaClass` instances obtained from ArchUnit in the first place. Apparently it hands out those in  uninitialized state.

**Comment by barthand** [comment]
Sure, here it is for you:
[moduliths-reactive-sample.zip](https://github.com/odrotbohm/moduliths/files/5957687/moduliths-reactive-sample.zip)

Just download it and run tests using `./mvnw test`. I was running it using AdoptOpenJDK 11.0.6.hs.

There are 3 tests out there, each created based on either docs or [presentation you gave on Moduliths](https://www.youtube.com/watch?v=Y4T55XMpP_Y):

* `ModulithsDemoApplicationTests` - tests Spring Boot context loading
* `ModularityTest` - purposed to tests modularity across modules (for sake of this example I stripped sample to single module)
* `DocumentingTest` - one that is failing (same code as in code sample pasted above)

I run the sample with [current code from main branch](https://github.com/odrotbohm/moduliths/tree/3ee2c7e3adc4f26e3be1abd29a1e385d13eed3b9) (with ArchUnit 0.16 in place), but still no luck, unfortunately -- same issue happens.

Hope this helps!

**Comment by odrotbohm** [comment]
Perfect, will take it from here. Thanks!

**Comment by odrotbohm** [comment]
Looks like this is caused by a special arrangement in declared and actually used types within ArchUnit. @codecholeric has a proposed fix for that in TNG/ArchUnit#530 but a workaround would be to add `org.springframework.web.reactive` into the framework packages to be explicitly scanned.

The aforementioned improvement in ArchUnit should actually make our explicit list of framework packages to include in the scanning obsolete. I'll create a separate issue for that.

**Comment by odrotbohm** [comment]
This is in place now in the current snapshots. The sample project provided builds fine with those used.
